catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergApiV1GenericResource.java

```diff
@@ -97,31 +110,95 @@ public IcebergConfigResponse getConfig(
         .build();
   }
 
-  // TODO inject some path parameters to identify the table in a secure way to then do the
-  //  access-check against the table and eventually sign the request.
-  //  The endpoint can be tweaked sing the S3_SIGNER_ENDPOINT property, signing to be
-  //  enabled via S3_REMOTE_SIGNING_ENABLED.
   @POST
   @Path("/v1/{prefix}/s3-sign/{identifier}")
   @Blocking
-  public IcebergS3SignResponse s3sign(
+  public Uni<Response> s3sign(
       IcebergS3SignRequest request,
       @PathParam("prefix") String prefix,
-      @PathParam("identifier") String identifier) {
-    ParsedReference ref = decodePrefix(prefix).parsedReference();
+      @PathParam("identifier") String identifier)
+      throws NessieNotFoundException {
 
-    URI uri = URI.create(request.uri());
+    DecodedPrefix decodedPrefix = decodePrefix(prefix);
+    ContentKey key = ContentKey.fromPathString(identifier);
+    String requestedS3Uri = S3Utils.asS3Location(request.uri());
+
+    return collectBasePrefixes(decodedPrefix, key)
+        .filter(requestedS3Uri::startsWith)
+        .map(ignored -> doRequestSign(request, decodedPrefix, key))
+        .toUni()
+        .replaceIfNullWith(() -> signingRequestDenied(request));
+  }
+
+  private Multi<String> collectBasePrefixes(DecodedPrefix decodedPrefix, ContentKey key)
+      throws NessieNotFoundException {
+    try {
+      ParsedReference ref = decodedPrefix.parsedReference();
+      ContentResponse contentResponse =
+          nessieApi
+              .getContent()
+              .refName(ref.name())
+              .hashOnRef(ref.hashWithRelativeSpec())
+              .getSingle(key);
+
+      // table exists: fetch snapshot, get additional locations
+      CompletionStage<SnapshotResponse> stage =
+          catalogService.retrieveSnapshot(
+              SnapshotReqParams.forSnapshotHttpReq(ref, "iceberg", null),
+              key,
+              new CatalogUriResolverImpl(uriInfo),
+              contentResponse.getContent().getType());
+      return Uni.createFrom()
+          .completionStage(stage)
+          .map(SnapshotResponse::nessieSnapshot)
+          .onItem()
+          .<String>transformToMulti(
+              snapshot ->
+                  Multi.createFrom()
+                      .emitter(
+                          e -> {
+                            e.emit(snapshot.icebergLocation());
+                            snapshot.additionalKnownLocations().forEach(e::emit);
+                            e.complete();
+                          }))
+          // allow the parent directory of each iceberg location,
+          // assuming the locations are of the form: <base-prefix>/<random-part>
+          .map(location -> location.substring(0, location.lastIndexOf('/') + 1));
```

Comment:
The location(s) is/are always absolute base paths. It is safe to assume that everything for a table is underneath one of these paths.
I.e. the correct way here would be to append a `/` (if not already present), but truncating would expose all tables in the same namaspace.

Reply:
I've hardened the logic a bit, let me know if that's better. Unfortunately we cannot just take the locations as-is.